src/main/java/dev/openfeature/sdk/EventProvider.java

```diff
@@ -16,21 +18,56 @@
  */
 public abstract class EventProvider implements FeatureProvider {
 
+    private ProviderState providerState = ProviderState.NOT_READY;
+
     /**
      * {@inheritDoc}
      */
     @Override
-    public abstract ProviderState getState();
+    public final ProviderState getState() {
+        return providerState;
+    }
+
+    @Override
+    public final void initialize(EvaluationContext evaluationContext) throws Exception {
+        try {
+            doInitialization(evaluationContext);
+            providerState = ProviderState.READY;
+        } catch (OpenFeatureError openFeatureError) {
+            if (ErrorCode.PROVIDER_FATAL.equals(openFeatureError.getErrorCode())) {
+                providerState = ProviderState.FATAL;
+            } else {
+                providerState = ProviderState.ERROR;
+            }
+            throw openFeatureError;
+        } catch (Exception e) {
+            providerState = ProviderState.ERROR;
+            throw new GeneralError(e);
+        }
+    }
+
+    protected void doInitialization(EvaluationContext evaluationContext) throws Exception {
+        // Intentionally left blank, to be implemented by inheritors
+    }
```

Comment:
This implementation is logically correct, but this won't work retroactively with existing implementations (many of which can be found [here](https://github.com/open-feature/dotnet-sdk/pull/276), though there's others maintained by vendors, etc).

We need to do this in a non-breaking way, without breaking any existing contracts or behvaiors. In [C#,](https://github.com/open-feature/dotnet-sdk/pull/276) I accomplished this by creating a new package-private ProviderState field on the provider which the SDK updates after it has successfully (or unsuccessfully) run the `initialize` method, and basically ignoring the previous provider state field.

This allows existing providers to work without breaking changes.

Reply:
@toddbaert I wonder if there can be some additional automated step to see if the _sdk-contrib_ repository providers are not breaking, possibly as warning / non-mandatory. There is the _InMemoryProvider_ on this repository which is a good sanity, maybe checking the other providers as well can help. Or is too complex ? What do you think ?